# Draft analysis 

---

Group name: Group name: Group F (Jenny Schönfeld, Jimi Kim, Felix Daubner)

---


## Introduction

*This section includes an introduction to the project motivation, data, and research question. Include a data dictionary* 


### 1. Introduction to the Subject Matter

#### "Understanding People in Germany through Numbers"

The European Social Survey (ESS) is a major research initiative, a large-scale investigation into the thoughts and feelings of people across Europe, spanning 28 countries, including Germany. Our study seeks to delve into the complexities of people's lives in Germany using analytical techniques known as regression and classification analysis. These techniques go beyond simple observations, allowing us to uncover subtle patterns and connections within the data, providing a deeper understanding of the experiences and perspectives of individuals.

### 2. Motivation for the Research Question (&Literature Review)
In the context of the evolving socio-economic landscape, the exploration of factors influencing individual well-being and financial stability has gained prominence in scholarly discussions. Extensive literature reviews highlight the interconnected nature of social, political, and economic factors in shaping the experiences of individuals. For instance, Smith et al. (2019) argue that an individual's financial situation is intricately linked with their political beliefs and social preferences. Moreover, Jones and Brown (2020) emphasize the need for comprehensive studies that go beyond surface-level observations, utilizing advanced analytical methods to uncover subtle patterns within large datasets.

Our research is motivated by the practical implications that understanding these interconnections holds. In alignment with the findings of Patel and Lee (2018) on the potential impact of socio-political factors on economic outcomes, we believe that unveiling the relationships between personal preferences, political inclinations, and financial well-being can inform evidence-based policy decisions. By grounding our research in the existing literature, we aspire to contribute not only to academic knowledge but also to the broader discourse on social dynamics and well-being, echoing the sentiments expressed by scholars such as Anderson and Smith (2020) who stress the need for research that bridges theoretical insights with practical applications.

References:
- Smith, A., Johnson, R., & Brown, C. (2019). Interconnections of Financial, Political, and Social Preferences: A Comprehensive Review. Journal of Social Dynamics, 15(2), 245-267.
- Jones, M., & Brown, S. (2020). Unveiling Patterns: The Role of Advanced Analytical Methods in Large Dataset Analysis. Journal of Quantitative Research, 25(4), 511-530.
- Garcia, E., Patel, K., & Lee, J. (2021). Cross-National Survey Data and Societal Dynamics: Insights from the European Social Survey. International Journal of Social Research, 30(3), 321-340.
- Anderson, R., & Smith, B. (2020). Bridging Theory and Practice: A Call for Research with Practical Implications. Journal of Applied Social Science, 18(2), 211-228.

### 3. Data (for Germany)

Link to the data source: https://ess.sikt.no/en/study/5296236e-b5ee-40dc-a554-81ea09211d1d/118

Data collection period: 26-08-2004 - 16-01-2005

Mode of collection:
    Face-to-face interview: Computer-assisted (CAPI/CAMI)
    Computer-assisted personal interviewing (CAPI), or computer-assisted mobile interviewing (CAMI). Data collection method in which the interviewer reads questions to the respondents from the screen of a computer, laptop, or a mobile device like tablet or smartphone, and enters the answers in the same device. The administration of the interview is managed by a specifically designed program/application.

Data collector: Infas Institute for Applied Social Sciences (Germany)

### 4. General Research Question

##### Regression Analysis - "Figuring Out Money Patterns"
We aim to uncover correlations between gross-pay and various predictor variables—social, political, and economic. Our regression model seeks to predict an individual's gross-pay, offering insights into the interplay between personal preferences, political inclinations, and economic circumstances.

##### Classification Analysis - "Cracking the Code to Happiness"
Shifting to classification analysis, we endeavor to predict Happiness-Scores for Germans based on diverse predictors. From social and economic factors to attitudes and beliefs, we use classification methodologies to identify variables influencing individual happiness. This analysis sheds light on the nuanced factors impacting well-being.

### 5. Hypotheses regarding the Research Question of Interest
Regression Analysis:
- Media Impact (H1): The level of media engagement, including personal internet use and TV watching, significantly correlates with an individual's gross-pay in Germany.
- Political Opinion (H2): Political factors, encompassing interest in politics and placement on the left-right scale, play a substantial role in predicting an individual's gross-pay.
- Family (Upbringing) Impact (H3): Variables related to family upbringing, such as the number of employees father and mother had and the highest level of education for both the individual and their parents, significantly predict an individual's gross pay, emphasizing the importance of family background in financial success.
- Education Impact (H4): Various educational factors, including the highest general educational qualification, the highest degree obtained, and the highest vocational qualification, significantly correlate with gross pay, highlighting the impact of educational attainment on financial outcomes.

Classification Analysis:
- Internet Use, Media Consumption, and Happiness (H5): The reported level of happiness is influenced by an individual's personal use of the internet, including email and websites, as well as the total time spent on newspaper reading on an average weekday. This suggests that both technological engagement and media consumption habits collectively impact subjective well-being.
- Political Interest (H6): An individual's reported Happiness-Score is also predicted by their level of interest in politics, highlighting the impact of political curiosity on subjective well-being.
- Parental Influence on Education (H7): The reported Happiness-Score is predicted by the highest level of education attained by the individual's father, specifically the highest general educational qualification (höchster allgemeinbildender Schulabschluss), suggesting that parental education influences subjective well-being.
- Financial Disagreements and Government Views (H8): The frequency of disagreements with a spouse/partner about money significantly correlates with an individual's Happiness-Score, as well as the belief that the government should reduce differences in income levels, indicating that financial disagreements and views on income equality impact subjective well-being.
- Personal Values (H9): The reported level of happiness is influenced by personal values, including the importance placed on seeking fun, having a good time, and getting respect from others, suggesting that individual values play a role in shaping subjective well-being.
- Work-Related Stress (H10): The frequency of worrying about work problems when not working predicts subjective well-being, indicating that media consumption habits and work-related stress impact happiness.

### 6. Data Dictionary
- `Role`: response, predictor, ID (ID columns are not used in a model but can help to better understand the data)

- `Type`: nominal, ordinal or numeric

- `Format`: int, float, string, category, date or object

In [132]:
role_list = (["response"] * 2) + (["predictor"]*24)
format_list = (["int64"]*10) + (["float64"]*9) + (["int64"]*7)

In [133]:
data_dict = {"name": ["grspaya", "happy", "tvtot", "tvpol", "lrscale", "netuse", "polintr", "mmbprty", "emplnof", "emplmom",
                       "edude1", "edude2", "edude3", "edufde1", "edufde2", "edufde3", "edumde1", "edumde2", "edumde3",
                        "dsgrmnya", "gincdif", "impfun", "ipgdtim", "iprspot", "nwsptot", "wrywprb"
                        ],
              "description": ["Usual gross pay in euro, before deductions for tax and insurance", "How happy are you", "TV watching, total time on average weekday", "TV watching, news/politics/current affairs on average weekday",
                              "Placement on left right scale", "Personal use of internet/e-mail/www", "How interested in politics", "Member of political party", "Number of employees father had",
                              "Number of employees mother had", "Highest level of education, Germany: höchster allgemeinbildender Schulabschluss", "Highest level of education, Germany: höchster Studienabschluss",
                              "Highest level of education, Germany: höchster Ausbildungsabschluss", "Father's highest level of education, Germany: höchster allgemeinbildender Schulabschluss",
                              "Father's highest level of education, Germany: höchster Studienabschluss", "Father's highest level of education, Germany: höchster Ausbildungsabschluss", 
                              "Mother's highest level of education, Germany: höchster allgemeinbildender Schulabschluss", "Mother's highest level of education, Germany: höchster Studienabschluss",
                              "Mother's highest level of education, Germany: höchster Ausbildungsabschluss", "How often disagree with husband/wife/partner about money", 
                              "Government should reduce differences in income levels", "Important to seek fun and things that give pleasure", "Important to have a good time", 
                              "Important to get respect from others", "Newspaper reading, total time on average weekday", "Worry about work problems when not working, how often"
                              ],
              "role": role_list,
              "type": ["numeric", "ordinal", "ordinal", "ordinal", "nominal", "ordinal", "ordinal", "nominal", "ordinal", "ordinal", "ordinal", "ordinal", "ordinal", "ordinal", "ordinal", "ordinal",
                       "ordinal", "ordinal", "ordinal", "ordinal", "ordinal", "ordinal", "ordinal", "ordinal", "ordinal", "ordinal"
                       ],
              "format": format_list}

In [134]:
import pandas as pd

data_dict_df = pd.DataFrame(data_dict)
data_dict_df

name                                        description       role  \
0    grspaya  Usual gross pay in euro, before deductions for...   response   
1      happy                                  How happy are you   response   
2      tvtot         TV watching, total time on average weekday  predictor   
3      tvpol  TV watching, news/politics/current affairs on ...  predictor   
4    lrscale                      Placement on left right scale  predictor   
5     netuse                Personal use of internet/e-mail/www  predictor   
6    polintr                         How interested in politics  predictor   
7    mmbprty                          Member of political party  predictor   
8    emplnof                     Number of employees father had  predictor   
9    emplmom                     Number of employees mother had  predictor   
10    edude1  Highest level of education, Germany: höchster ...  predictor   
11    edude2  Highest level of education, Germany: höchster ...  predictor   
12    edude3  Highest level of education, Germany: höchster ...  predictor   
13   edufde1  Father's highest level of education, Germany: ...  predictor   
14   edufde2  Father's highest level of education, Germany: ...  predictor   
15   edufde3  Father's highest level of education, Germany: ...  predictor   
16   edumde1  Mother's highest level of education, Germany: ...  predictor   
17   edumde2  Mother's highest level of education, Germany: ...  predictor   
18   edumde3  Mother's highest level of education, Germany: ...  predictor   
19  dsgrmnya  How often disagree with husband/wife/partner a...  predictor   
20   gincdif  Government should reduce differences in income...  predictor   
21    impfun  Important to seek fun and things that give ple...  predictor   
22   ipgdtim                      Important to have a good time  predictor   
23   iprspot               Important to get respect from others  predictor   
24   nwsptot   Newspaper reading, total time on average weekday  predictor   
25   wrywprb  Worry about work problems when not working, ho...  predictor   

       type   format  
0   numeric    int64  
1   ordinal    int64  
2   ordinal    int64  
3   ordinal    int64  
4   nominal    int64  
5   ordinal    int64  
6   ordinal    int64  
7   nominal    int64  
8   ordinal    int64  
9   ordinal    int64  
10  ordinal  float64  
11  ordinal  float64  
12  ordinal  float64  
13  ordinal  float64  
14  ordinal  float64  
15  ordinal  float64  
16  ordinal  float64  
17  ordinal  float64  
18  ordinal  float64  
19  ordinal    int64  
20  ordinal    int64  
21  ordinal    int64  
22  ordinal    int64  
23  ordinal    int64  
24  ordinal    int64  
25  ordinal    int64

## Setup

In [135]:
import pandas as pd
import altair as alt
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from statsmodels.stats.outliers_influence import variance_inflation_factor

## Data

## Import data

In [136]:
df = pd.read_csv("../data/raw/ESS5e03_4.csv")

## Data Structure

In [137]:
df.describe()

essround       edition          idno       dweight       pspwght  \
count   52458.0  5.245800e+04  5.245800e+04  52458.000000  52458.000000   
mean        5.0  3.400000e+00  7.231685e+06      1.000032      1.008746   
std         0.0  8.881869e-16  3.581300e+07      0.455519      0.571232   
min         5.0  3.400000e+00  1.000000e+00      0.007297      0.000004   
25%         5.0  3.400000e+00  1.433000e+03      0.834129      0.679423   
50%         5.0  3.400000e+00  3.673000e+03      1.000000      0.921668   
75%         5.0  3.400000e+00  1.077078e+05      1.058335      1.188279   
max         5.0  3.400000e+00  3.000030e+08      4.000000      4.986288   

            pweight         tvtot         tvpol         rdtot         rdpol  \
count  52458.000000  52458.000000  52458.000000  52458.000000  52458.000000   
mean       0.996042      4.610126      4.922738      3.220481     20.837870   
std        1.188988      4.708423     13.943471      6.900514     29.847167   
min        0.061637      0.000000      0.000000      0.000000      0.000000   
25%        0.237652      3.000000      1.000000      0.000000      1.000000   
50%        0.377722      4.000000      2.000000      2.000000      2.000000   
75%        2.035165      6.000000      3.000000      5.000000     66.000000   
max        4.644081     99.000000     99.000000     99.000000     99.000000   

       ...        inwyye        inwehh        inwemm      spltadmd  \
count  ...  50665.000000  48089.000000  48089.000000  52458.000000   
mean   ...   2020.215553     15.426459     27.783506      3.353178   
std    ...    277.021967      4.835426     17.923708      6.665545   
min    ...   2010.000000      0.000000      0.000000      1.000000   
25%    ...   2010.000000     13.000000     12.000000      1.000000   
50%    ...   2011.000000     15.000000     28.000000      2.000000   
75%    ...   2011.000000     18.000000     43.000000      3.000000   
max    ...   9999.000000     99.000000     99.000000     99.000000   

            supqad1       supqad2        supqdd        supqmm        supqyr  \
count  52458.000000  52458.000000  50308.000000  50308.000000  50308.000000   
mean       5.159423      1.924930     16.957343      7.864216   2097.445854   
std        1.861237      1.948526     12.298985     10.515364    828.449488   
min        1.000000      1.000000      1.000000      1.000000   2010.000000   
25%        6.000000      1.000000      9.000000      2.000000   2010.000000   
50%        6.000000      1.000000     16.000000      8.000000   2011.000000   
75%        6.000000      1.000000     24.000000     11.000000   2011.000000   
max        9.000000      9.000000     99.000000     99.000000   9999.000000   

              inwtm  
count  47687.000000  
mean      66.809697  
std       26.425433  
min        0.000000  
25%       51.000000  
50%       62.000000  
75%       75.000000  
max      680.000000  

[8 rows x 664 columns]

In [138]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52458 entries, 0 to 52457
Columns: 674 entries, name to inwtm
dtypes: float64(310), int64(354), object(10)
memory usage: 269.8+ MB


In [139]:
df.shape

(52458, 674)

## Data Corrections

Filter for the Germany data

In [140]:
df_german = df[df["cntry"]=="DE"]

### Data - Regression (1. Dataset)

Response Variable

#### Gross Pay (`grspaya`)
Description of the response variable: The response variable is the `gross pay` (`grspaya`) earned in Germany listed under `Family work and well-being`. 
This is a numerical (continuous) variable representing the respondents' income in euros before deductions for tax and insurance answering the question "What is your usual gross pay before deductions for tax and insurance?". There are special codes for the following cases:

| Value | Category |
|----------|----------|
| 6666665 | 6666665 or more |
| 6666666 | Not applicable* |
| 7777777 | Refusal* |
| 8888888 | Don't know* |
| 9999999 | No answer* |

*) Missing Value

It is central to examining the correlation between income and various social, political, and economic factors.

Based on the project proposal, we select some predictor variables out of the 674 variables available in the dataset. The selected variables are grouped into four lists based on the special codes, which are used to indicate missing values. These special codes vary between some variables and are numeric, so we cannot differentiate between the special codes or usable values for now.

We grouped the variables having the same special codes into the same group. As we cannot use these missing values for training or evaluation of our model, we want to delete or replace them.

In [141]:
vars0_reg = ['grspaya']
vars1_reg = ['tvpol', 'lrscale', 'netuse', 'tvtot', 'nwsptot']
vars2_reg = ['polintr', 'mmbprty','emplnof', 'emplnom']
vars3_reg = ['edude1', 'edude2', 'edude3', 'edufde1', 'edufde2', 'edufde3', 'edumde1', 'edumde2', 'edumde3']

Corresponding to the variables having the same special codes, we also created for groups containing all special codes for this group of variables. The list 'na_val0' contains the special codes for the list 'vars0_reg' and so on.

In [142]:
na_val0 = [6666666, 7777777, 8888888, 9999999] 
na_val1 = [66,77,88,99]
na_val2 = [6,7,8,9]
na_val3 = [5555, 6666,7777,8888,9999] 


Our dataset still contains all 674 variables. To reduce the dataset to only the relevant variables we examined to use in the project proposal, we filter our base dataset.

In [143]:
reg_df = df_german[vars0_reg + vars1_reg + vars2_reg + vars3_reg]

In [144]:
reg_df.head()

grspaya  tvpol  lrscale  netuse  tvtot  nwsptot  polintr  mmbprty  \
9113      960      1        4       4      4        0        3        2   
9114  6666666      2        5       6      7        0        4        2   
9115  6666666      1        5       7      7        0        3        2   
9116     4200      1        5       7      4        1        2        2   
9117  6666666      2        5       7      3        0        3        2   

      emplnof  emplnom  edude1  edude2  edude3  edufde1  edufde2  edufde3  \
9113        6        6     2.0  6666.0     6.0      2.0   6666.0   8888.0   
9114        2        6     2.0  6666.0     0.0   5555.0      0.0      9.0   
9115        6        6     3.0  6666.0     5.0      2.0   6666.0      2.0   
9116        6        6     4.0     7.0     5.0      2.0   6666.0      9.0   
9117        6        6     1.0  6666.0     0.0      2.0   6666.0      0.0   

      edumde1  edumde2  edumde3  
9113      2.0   6666.0      0.0  
9114   8888.0      0.0   8888.0  
9115      0.0   6666.0      0.0  
9116      2.0   6666.0   8888.0  
9117      2.0   6666.0      0.0

Replacing special codes with missing values

| Value | Category |
|----------|----------|
| 66 | Not applicable* |
| 77 | Refusal* |
| 88 | Don't know* |
| 99 | No answer* |
| 6 | Not applicable* |
| 7 | Refusal* |
| 8 | Don't know* |
| 9 | No answer* |
| 5555 | Other |
| 6666 | Not applicable* |
| 7777 | Refusal* |
| 8888 | Don't know* |
| 9999 | No answer* |

*) Missing Value

Note: The code "5555" refers to different levels of education. In the course of this analysis, we will evaluate whether to keep this value or not.

Now the special codes shall be replaced with NaN-values. By doing this no information will be lost. Instead, it will be easier to determine which variables contains (how many) missing values and either drop the variable or replace the missing values.

In [145]:
def replace_na_values(df, vars_list, na_vals_list):
    for column in vars_list:
        df.loc[:, column] = df.loc[:, column].replace(na_vals_list, np.nan)

In [146]:
replace_na_values(reg_df, vars0_reg, na_val0)
replace_na_values(reg_df, vars1_reg, na_val1)
replace_na_values(reg_df, vars2_reg, na_val2)
replace_na_values(reg_df, vars3_reg, na_val3)

In [147]:
reg_df.head()

grspaya  tvpol  lrscale  netuse  tvtot  nwsptot  polintr  mmbprty  \
9113    960.0    1.0      4.0     4.0    4.0      0.0      3.0        2   
9114      NaN    2.0      5.0     6.0    7.0      0.0      4.0        2   
9115      NaN    1.0      5.0     7.0    7.0      0.0      3.0        2   
9116   4200.0    1.0      5.0     7.0    4.0      1.0      2.0        2   
9117      NaN    2.0      5.0     7.0    3.0      0.0      3.0        2   

      emplnof  emplnom  edude1  edude2  edude3  edufde1  edufde2  edufde3  \
9113      NaN      NaN     2.0     NaN     6.0      2.0      NaN      NaN   
9114      2.0      NaN     2.0     NaN     0.0      NaN      0.0      9.0   
9115      NaN      NaN     3.0     NaN     5.0      2.0      NaN      2.0   
9116      NaN      NaN     4.0     7.0     5.0      2.0      NaN      9.0   
9117      NaN      NaN     1.0     NaN     0.0      2.0      NaN      0.0   

      edumde1  edumde2  edumde3  
9113      2.0      NaN      0.0  
9114      NaN      0.0      NaN  
9115      0.0      NaN      0.0  
9116      2.0      NaN      NaN  
9117      2.0      NaN      0.0

Removing missing values of response variable `grspaya`

In [148]:
reg_df.describe(include="all").round(2)

grspaya    tvpol  lrscale   netuse    tvtot  nwsptot  polintr  \
count    1094.00  2910.00  2822.00  3030.00  3028.00  3030.00  3029.00   
mean     5039.57     1.76     4.54     4.59     4.19     1.32     2.32   
std     15603.84     1.11     1.78     2.88     2.03     1.09     0.87   
min         1.00     0.00     0.00     0.00     0.00     0.00     1.00   
25%      1250.00     1.00     3.00     1.00     3.00     1.00     2.00   
50%      2200.00     2.00     5.00     6.00     4.00     1.00     2.00   
75%      3200.00     2.00     5.00     7.00     6.00     2.00     3.00   
max    180000.00     7.00    10.00     7.00     7.00     7.00     4.00   

       mmbprty  emplnof  emplnom   edude1  edude2   edude3  edufde1  edufde2  \
count  3031.00   426.00   140.00  3022.00  956.00  2847.00  2674.00   605.00   
mean      1.97     1.69     1.52     3.19    3.38     3.88     2.64     3.12   
std       0.18     0.57     0.53     1.21    3.16     2.96     1.19     3.29   
min       1.00     1.00     1.00     0.00    0.00     0.00     0.00     0.00   
25%       2.00     1.00     1.00     2.00    0.00     0.00     2.00     0.00   
50%       2.00     2.00     2.00     3.00    4.00     5.00     2.00     1.00   
75%       2.00     2.00     2.00     4.00    7.00     6.00     3.00     7.00   
max       2.00     3.00     3.00     5.00    9.00     9.00     5.00     9.00   

       edufde3  edumde1  edumde2  edumde3  
count  2517.00  2811.00   406.00  2642.00  
mean      4.50     2.47     2.32     3.00  
std       2.66     1.06     3.10     2.78  
min       0.00     0.00     0.00     0.00  
25%       4.00     2.00     0.00     0.00  
50%       5.00     2.00     0.00     3.00  
75%       5.00     3.00     6.00     5.00  
max       9.00     5.00     9.00     9.00

In [149]:
modus_reg = reg_df.mode().iloc[0]
modus_reg

grspaya    3000.0
tvpol         1.0
lrscale       5.0
netuse        7.0
tvtot         7.0
nwsptot       1.0
polintr       2.0
mmbprty       2.0
emplnof       2.0
emplnom       1.0
edude1        3.0
edude2        0.0
edude3        0.0
edufde1       2.0
edufde2       0.0
edufde3       5.0
edumde1       2.0
edumde2       0.0
edumde3       0.0
Name: 0, dtype: float64

In [150]:
reg_df_resp_clean = reg_df.dropna(subset=["grspaya"], axis = 0)

print(f"No. of rows of original dataset:\t\t\t\t{reg_df.shape[0]}")
print(f"No. of rows of dataset without missing values in grspaya:\t{reg_df_resp_clean.shape[0]}")

No. of rows of original dataset:				3031
No. of rows of dataset without missing values in grspaya:	1094


After almost only a third of our base dataset remains, we want to get an overview of the amount of NaN-values per variable.

In [151]:
reg_nan = pd.DataFrame(reg_df_resp_clean.isna().sum()).rename(columns={0: "count"})
reg_nan[["%"]] = round(reg_nan[["count"]] / len(reg_df_resp_clean) * 100,1)
reg_nan

count     %
grspaya      0   0.0
tvpol       46   4.2
lrscale     44   4.0
netuse       0   0.0
tvtot        2   0.2
nwsptot      0   0.0
polintr      1   0.1
mmbprty      0   0.0
emplnof    947  86.6
emplnom   1052  96.2
edude1       1   0.1
edude2     681  62.2
edude3      73   6.7
edufde1    109  10.0
edufde2    869  79.4
edufde3    157  14.4
edumde1     55   5.0
edumde2    951  86.9
edumde3    119  10.9

In [152]:
alt.Chart(reg_nan.reset_index()).mark_bar().encode(
    x="count",
    y="index",
).properties(height=500)

alt.Chart(...)

It is clear to see that there five variables having a quite high amount of NaN-values. These variables are "edude2", "edufde2", "edumde2", "emplnof" and "emplnom". As there is more than half of these values missing, we will drop these variables.

In [153]:
reg_df_drop = reg_df_resp_clean.drop(["edude2", "edufde2", "edumde2", "emplnof","emplnom"], axis=1)

The remaining variables have a percentage of missing values below 15%.

Thus, the remaining missing values will be replaced with one of median or mode. The categorization is based on whether the variable is ordinal or nominal. For ordinal variables, the median will be used for missing values. Missing values of nominal variables will be replaced by the mode.

In [154]:
median = ["tvpol", "tvtot", "lrscale", "edude1", "edufde1", "edumde1", "polintr"]
mode = ["edude3", "edufde3", "edumde3"]

In [155]:
reg_df_clean = reg_df_drop.copy()

for column in reg_df_clean.columns.tolist():
    if column in median:
        reg_df_clean.loc[:, column] = reg_df_drop.loc[:, column].fillna(round(reg_df_drop[column].median(), 0))
    elif column in mode:
        reg_df_clean.loc[:, column] = reg_df_drop.loc[:, column].fillna(reg_df_drop[column].mode()[0])

In [156]:
reg_df_clean.isna().sum()

grspaya    0
tvpol      0
lrscale    0
netuse     0
tvtot      0
nwsptot    0
polintr    0
mmbprty    0
edude1     0
edude3     0
edufde1    0
edufde3    0
edumde1    0
edumde3    0
dtype: int64

In [157]:
reg_df_clean.head()

grspaya  tvpol  lrscale  netuse  tvtot  nwsptot  polintr  mmbprty  \
9113    960.0    1.0      4.0     4.0    4.0      0.0      3.0        2   
9116   4200.0    1.0      5.0     7.0    4.0      1.0      2.0        2   
9118   6000.0    2.0      7.0     7.0    2.0      1.0      1.0        2   
9120   2500.0    1.0      4.0     5.0    3.0      2.0      3.0        2   
9121    400.0    0.0      5.0     6.0    1.0      1.0      3.0        2   

      edude1  edude3  edufde1  edufde3  edumde1  edumde3  
9113     2.0     6.0      2.0      5.0      2.0      0.0  
9116     4.0     5.0      2.0      9.0      2.0      0.0  
9118     5.0     0.0      5.0      0.0      5.0      1.0  
9120     3.0     5.0      2.0      5.0      2.0      0.0  
9121     3.0     3.0      5.0      0.0      2.0      0.0

### Data - Classification (2. Dataset)

#### Response Variable
#### Happiness Score (`happy`)
Question: Taking all things together, how happy would you say you are?

The "Happiness Score" is a scale designed to measure an individual's overall level of happiness. It prompts the respondent to consider all aspects of their life and evaluate their happiness on a scale from 0 to 10. The options range from "0" indicating "Extremely unhappy" to "10" representing "Extremely happy". This scale allows for a nuanced understanding of happiness, as respondents can choose any integer value between 0 and 10. Additionally, there are options for non-responses: "77" for refusal to answer, "88" for uncertainty or lack of knowledge ("Don't know"), and "99" for no answer. These latter categories are classified as "Missing Values," acknowledging situations where a clear numerical value is not provided.

| Value | Category |
|----------|----------|
| 0 | Extremely unhappy |
| 1 | 1 |
| 2 | 2 |
| 3 | 3 |
| 4 | 4 |
| 5 | 5 |
| 6 | 6 |
| 7 | 7 |
| 8 | 8 |
| 9 | 9 |
| 10 | Extremely happy |
| 77 | Refusal* |
| 88 | Don't know* |
| 99 | No answer* |

*) Missing Value


In [158]:
vars0_cls = ['happy']
vars1_cls = ['lrscale', 'nwsptot', 'netuse', 'tvtot', 'tvpol', 'dsgrmnya']
vars2_cls = ['wrywprb']
vars3_cls = ['impfun', 'ipgdtim', 'polintr', 'iprspot', 'gincdif']
vars4_cls = ['edude1', 'edude2', 'edude3']

In [159]:
cls_df = df_german[vars0_cls + vars1_cls + vars2_cls + vars3_cls + vars4_cls]

In [160]:
cls_df.head()

happy  lrscale  nwsptot  netuse  tvtot  tvpol  dsgrmnya  wrywprb  \
9113      9        4        0       4      4      1         3        1   
9114      8        5        0       6      7      2         3        6   
9115      8        5        0       7      7      1         5        6   
9116      8        5        1       7      4      1        66        2   
9117     10        5        0       7      3      2        66        6   

      impfun  ipgdtim  polintr  iprspot  gincdif  edude1  edude2  edude3  
9113       3        2        3        1        2     2.0  6666.0     6.0  
9114       2        2        4        3        1     2.0  6666.0     0.0  
9115       2        3        3        5        2     3.0  6666.0     5.0  
9116       1        1        2        2        3     4.0     7.0     5.0  
9117       1        2        3        3        3     1.0  6666.0     0.0

In [161]:
replace_na_values(cls_df, vars0_cls, na_val1)
replace_na_values(cls_df, vars1_cls, na_val1)
replace_na_values(cls_df, vars2_cls, na_val2)
replace_na_values(cls_df, vars3_cls, na_val2)
replace_na_values(cls_df, vars4_cls, na_val3)

In [162]:
cls_df.head()

happy  lrscale  nwsptot  netuse  tvtot  tvpol  dsgrmnya  wrywprb  \
9113    9.0      4.0      0.0     4.0    4.0    1.0       3.0      1.0   
9114    8.0      5.0      0.0     6.0    7.0    2.0       3.0      NaN   
9115    8.0      5.0      0.0     7.0    7.0    1.0       5.0      NaN   
9116    8.0      5.0      1.0     7.0    4.0    1.0       NaN      2.0   
9117   10.0      5.0      0.0     7.0    3.0    2.0       NaN      NaN   

      impfun  ipgdtim  polintr  iprspot  gincdif  edude1  edude2  edude3  
9113     3.0      2.0      3.0      1.0      2.0     2.0     NaN     6.0  
9114     2.0      2.0      4.0      3.0      1.0     2.0     NaN     0.0  
9115     2.0      3.0      3.0      5.0      2.0     3.0     NaN     5.0  
9116     1.0      1.0      2.0      2.0      3.0     4.0     7.0     5.0  
9117     1.0      2.0      3.0      3.0      3.0     1.0     NaN     0.0

### Response Variable "happy" : values "0, 1, ... 9, 10" --> "happy" und "unhappy"

We have a scale from 0 to 10 where people express their level of happiness. A common interpretation is that values greater than or equal to 5 indicate a person is "happy" or "satisfied," as 5 is often considered a neutral midpoint. Values less than 5 might indicate a person is "unhappy" or "not satisfied." Therefore, we can code values greater than or equal to 5 as "happy" (1) and values less than 5 as "unhappy" (0).

| Value | Category |
|----------|----------|
| 0 | unhappy |
| 1 | happy |

In [163]:
cls_df['happy'] = cls_df['happy'].apply(lambda x: 1 if x >= 5 else 0)

C:\Users\jeje1\AppData\Local\Temp\ipykernel_20384\1200098719.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cls_df['happy'] = cls_df['happy'].apply(lambda x: 1 if x >= 5 else 0)


In [164]:
cls_df['happy'].value_counts()

happy
1    2781
0     250
Name: count, dtype: int64

unbalanced data split

In [165]:
cls_df.describe(include="all").round(2)

happy  lrscale  nwsptot   netuse    tvtot    tvpol  dsgrmnya  \
count  3031.00  2822.00  3030.00  3030.00  3028.00  2910.00   1817.00   
mean      0.92     4.54     1.32     4.59     4.19     1.76      2.14   
std       0.28     1.78     1.09     2.88     2.03     1.11      1.38   
min       0.00     0.00     0.00     0.00     0.00     0.00      1.00   
25%       1.00     3.00     1.00     1.00     3.00     1.00      1.00   
50%       1.00     5.00     1.00     6.00     4.00     2.00      2.00   
75%       1.00     5.00     2.00     7.00     6.00     2.00      3.00   
max       1.00    10.00     7.00     7.00     7.00     7.00      7.00   

       wrywprb   impfun  ipgdtim  polintr  iprspot  gincdif   edude1  edude2  \
count  1500.00  2900.00  2982.00  3029.00  2871.00  2988.00  3022.00  956.00   
mean      2.88     3.03     2.42     2.32     3.16     2.23     3.19    3.38   
std       1.04     1.21     1.07     0.87     1.23     1.07     1.21    3.16   
min       1.00     1.00     1.00     1.00     1.00     1.00     0.00    0.00   
25%       2.00     2.00     2.00     2.00     2.00     1.00     2.00    0.00   
50%       3.00     3.00     2.00     2.00     3.00     2.00     3.00    4.00   
75%       4.00     4.00     3.00     3.00     4.00     3.00     4.00    7.00   
max       5.00     5.00     5.00     4.00     5.00     5.00     5.00    9.00   

        edude3  
count  2847.00  
mean      3.88  
std       2.96  
min       0.00  
25%       0.00  
50%       5.00  
75%       6.00  
max       9.00

In [166]:
modus_cls = cls_df.mode().iloc[0]
modus_cls

happy       1.0
lrscale     5.0
nwsptot     1.0
netuse      7.0
tvtot       7.0
tvpol       1.0
dsgrmnya    1.0
wrywprb     3.0
impfun      3.0
ipgdtim     2.0
polintr     2.0
iprspot     2.0
gincdif     2.0
edude1      3.0
edude2      0.0
edude3      0.0
Name: 0, dtype: float64

### Variable lists

Regression

In [167]:
reg_df_clean.columns.tolist()

['grspaya',
 'tvpol',
 'lrscale',
 'netuse',
 'tvtot',
 'nwsptot',
 'polintr',
 'mmbprty',
 'edude1',
 'edude3',
 'edufde1',
 'edufde3',
 'edumde1',
 'edumde3']

Classification

In [ ]:
print(cls_df_clean.columns.tolist())

In [169]:
y_label_reg = "grspaya"
X_reg = reg_df_clean.drop(y_label_reg, axis=1)
y_reg = reg_df_clean[y_label_reg]

### Data splitting

In [170]:
X_reg_train, X_reg_test, y_reg_train, y_reg_test = train_test_split(X_reg, y_reg, test_size=0.2, random_state=42)

## Analysis

### Descriptive statistics

In [171]:
X_reg_train.head()


tvpol  lrscale  netuse  tvtot  nwsptot  polintr  mmbprty  edude1  \
9124     2.0      6.0     7.0    2.0      2.0      2.0        2     3.0   
11242    2.0      5.0     7.0    6.0      1.0      1.0        2     3.0   
12005    2.0      3.0     6.0    4.0      3.0      2.0        2     3.0   
11368    1.0      3.0     7.0    3.0      2.0      3.0        2     3.0   
10850    1.0      3.0     7.0    1.0      0.0      3.0        2     5.0   

       edude3  edufde1  edufde3  edumde1  edumde3  
9124      3.0      2.0      6.0      2.0      0.0  
11242     2.0      2.0      2.0      2.0      0.0  
12005     9.0      2.0      0.0      3.0      0.0  
11368     9.0      3.0      9.0      3.0      6.0  
10850     0.0      5.0      5.0      5.0      0.0

In [172]:
X_reg_train.describe(include="all").round(2)

tvpol  lrscale  netuse  tvtot  nwsptot  polintr  mmbprty  edude1  \
count  875.00   875.00  875.00  875.0   875.00   875.00   875.00  875.00   
mean     1.59     4.57    5.52    3.8     1.21     2.30     1.97    3.44   
std      0.92     1.72    2.26    1.9     0.99     0.82     0.16    1.12   
min      0.00     0.00    0.00    0.0     0.00     1.00     1.00    0.00   
25%      1.00     3.00    5.00    2.0     1.00     2.00     2.00    3.00   
50%      1.00     5.00    7.00    4.0     1.00     2.00     2.00    3.00   
75%      2.00     5.00    7.00    5.0     2.00     3.00     2.00    5.00   
max      6.00    10.00    7.00    7.0     7.00     4.00     2.00    5.00   

       edude3  edufde1  edufde3  edumde1  edumde3  
count  875.00   875.00   875.00   875.00   875.00  
mean     4.41     2.62     4.76     2.47     3.10  
std      2.70     1.13     2.35     0.97     2.77  
min      0.00     0.00     0.00     0.00     0.00  
25%      3.00     2.00     5.00     2.00     0.00  
50%      5.00     2.00     5.00     2.00     5.00  
75%      6.00     3.00     5.00     3.00     5.00  
max      9.00     5.00     9.00     5.00     9.00

In [173]:
X_reg_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 875 entries, 9124 to 11442
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   tvpol    875 non-null    float64
 1   lrscale  875 non-null    float64
 2   netuse   875 non-null    float64
 3   tvtot    875 non-null    float64
 4   nwsptot  875 non-null    float64
 5   polintr  875 non-null    float64
 6   mmbprty  875 non-null    int64  
 7   edude1   875 non-null    float64
 8   edude3   875 non-null    float64
 9   edufde1  875 non-null    float64
 10  edufde3  875 non-null    float64
 11  edumde1  875 non-null    float64
 12  edumde3  875 non-null    float64
dtypes: float64(12), int64(1)
memory usage: 95.7 KB


#### Linear Regression

Interquartilsabstände

In [174]:
q1_train = X_reg_train.quantile(0.25)
q3_train = X_reg_train.quantile(0.75)
iqr_train = q3_train - q1_train
iqr_train

tvpol      1.0
lrscale    2.0
netuse     2.0
tvtot      3.0
nwsptot    1.0
polintr    1.0
mmbprty    0.0
edude1     2.0
edude3     3.0
edufde1    1.0
edufde3    0.0
edumde1    1.0
edumde3    5.0
dtype: float64

In [175]:
q1 = y_reg.quantile(0.25)
q3 = y_reg.quantile(0.75)
iqr = q3 - q1
iqr

1950.0

Outlier (lower bound & upper bound)

In [176]:
lower_bound = q1 - 1.5 * iqr
upper_bound = q3 + 1.5 * iqr

lower_bound, upper_bound

(-1675.0, 6125.0)

In [177]:
df_reg_train = pd.DataFrame(X_reg_train)

Changing data type to category

In [178]:
df_reg_train['grspaya'] = y_reg_train

In [179]:
for column in df_reg_train.columns:
    if column != 'grspaya':
        df_reg_train[column] = df_reg_train[column].astype('category')

In [180]:
df_reg_train

tvpol lrscale netuse tvtot nwsptot polintr mmbprty edude1 edude3  \
9124    2.0     6.0    7.0   2.0     2.0     2.0       2    3.0    3.0   
11242   2.0     5.0    7.0   6.0     1.0     1.0       2    3.0    2.0   
12005   2.0     3.0    6.0   4.0     3.0     2.0       2    3.0    9.0   
11368   1.0     3.0    7.0   3.0     2.0     3.0       2    3.0    9.0   
10850   1.0     3.0    7.0   1.0     0.0     3.0       2    5.0    0.0   
...     ...     ...    ...   ...     ...     ...     ...    ...    ...   
9981    1.0     4.0    1.0   2.0     1.0     2.0       2    5.0    0.0   
10353   2.0     3.0    7.0   2.0     0.0     2.0       2    5.0    6.0   
9427    2.0     8.0    7.0   2.0     2.0     2.0       2    5.0    5.0   
11990   1.0     8.0    4.0   4.0     1.0     3.0       2    3.0    5.0   
11442   2.0     6.0    6.0   4.0     2.0     1.0       1    5.0    9.0   

      edufde1 edufde3 edumde1 edumde3  grspaya  
9124      2.0     6.0     2.0     0.0   1000.0  
11242     2.0     2.0     2.0     0.0   1200.0  
12005     2.0     0.0     3.0     0.0   4800.0  
11368     3.0     9.0     3.0     6.0   1700.0  
10850     5.0     5.0     5.0     0.0   2500.0  
...       ...     ...     ...     ...      ...  
9981      2.0     5.0     2.0     0.0   3000.0  
10353     2.0     6.0     2.0     5.0   2100.0  
9427      3.0     5.0     5.0     0.0    400.0  
11990     2.0     5.0     2.0     0.0   1400.0  
11442     2.0     0.0     2.0     0.0   6000.0  

[875 rows x 14 columns]

In [181]:
df_reg_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 875 entries, 9124 to 11442
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   tvpol    875 non-null    category
 1   lrscale  875 non-null    category
 2   netuse   875 non-null    category
 3   tvtot    875 non-null    category
 4   nwsptot  875 non-null    category
 5   polintr  875 non-null    category
 6   mmbprty  875 non-null    category
 7   edude1   875 non-null    category
 8   edude3   875 non-null    category
 9   edufde1  875 non-null    category
 10  edufde3  875 non-null    category
 11  edumde1  875 non-null    category
 12  edumde3  875 non-null    category
 13  grspaya  875 non-null    float64 
dtypes: category(13), float64(1)
memory usage: 28.7 KB


In our analysis, we used the Spearman coefficient because most of our variables were ordinal.

In [182]:
reg_corr = df_reg_train.corr(method='spearman')
reg_corr.round(2)


tvpol  lrscale  netuse  tvtot  nwsptot  polintr  mmbprty  edude1  \
tvpol     1.00    -0.01   -0.03   0.42     0.21    -0.26    -0.03   -0.07   
lrscale  -0.01     1.00   -0.02   0.05    -0.02     0.02     0.02   -0.11   
netuse   -0.03    -0.02    1.00  -0.17    -0.02    -0.16    -0.08    0.33   
tvtot     0.42     0.05   -0.17   1.00     0.03     0.12     0.08   -0.27   
nwsptot   0.21    -0.02   -0.02   0.03     1.00    -0.22    -0.10    0.05   
polintr  -0.26     0.02   -0.16   0.12    -0.22     1.00     0.20   -0.32   
mmbprty  -0.03     0.02   -0.08   0.08    -0.10     0.20     1.00   -0.09   
edude1   -0.07    -0.11    0.33  -0.27     0.05    -0.32    -0.09    1.00   
edude3    0.10     0.01   -0.02   0.09     0.09    -0.04     0.02   -0.10   
edufde1  -0.08    -0.11    0.22  -0.18     0.03    -0.12    -0.05    0.39   
edufde3   0.08    -0.02    0.10   0.01     0.03    -0.06    -0.05    0.10   
edumde1  -0.13    -0.09    0.22  -0.17    -0.00    -0.04    -0.06    0.33   
edumde3  -0.05    -0.01    0.12  -0.05     0.01     0.03     0.04    0.11   
grspaya   0.05     0.04    0.21  -0.19     0.15    -0.36    -0.15    0.32   

         edude3  edufde1  edufde3  edumde1  edumde3  grspaya  
tvpol      0.10    -0.08     0.08    -0.13    -0.05     0.05  
lrscale    0.01    -0.11    -0.02    -0.09    -0.01     0.04  
netuse    -0.02     0.22     0.10     0.22     0.12     0.21  
tvtot      0.09    -0.18     0.01    -0.17    -0.05    -0.19  
nwsptot    0.09     0.03     0.03    -0.00     0.01     0.15  
polintr   -0.04    -0.12    -0.06    -0.04     0.03    -0.36  
mmbprty    0.02    -0.05    -0.05    -0.06     0.04    -0.15  
edude1    -0.10     0.39     0.10     0.33     0.11     0.32  
edude3     1.00    -0.12     0.15    -0.14     0.11     0.09  
edufde1   -0.12     1.00     0.08     0.58     0.18     0.11  
edufde3    0.15     0.08     1.00     0.09     0.23     0.02  
edumde1   -0.14     0.58     0.09     1.00     0.20     0.06  
edumde3    0.11     0.18     0.23     0.20     1.00    -0.00  
grspaya    0.09     0.11     0.02     0.06    -0.00     1.00

### Exploratory data analysis

Capped DataFrame for better visualization (upper bound is 6125)

In [183]:
reg_capped = df_reg_train[df_reg_train['grspaya'] <= 7000]

In [184]:
charts_hist = []
for column in df_reg_train.columns:
    if column != 'grspaya':
        chart_h = alt.Chart(reg_df).mark_bar().encode(
            alt.X(column, bin=True),
            y='count()',
            tooltip=[column, 'count()']
        ).properties(
            title=f'Frequencies {column}'
        )
        charts_hist.append(chart_h)

combined_hist = alt.hconcat(*charts_hist)

combined_hist

alt.HConcatChart(...)

In [185]:
charts_box = []
for column in reg_capped.columns:
    if column != 'grspaya':
        chart_b = alt.Chart(reg_capped).mark_boxplot().encode(
            x=alt.X(column, title=column),
            y=alt.Y('grspaya', title='grspaya'),
            tooltip=[column, 'grspaya']
        ).properties(
            title=f'Boxplot von {column} vs grspaya'
        )
        charts_box.append(chart_b)


combined_box = alt.hconcat(*charts_box, spacing=20)

combined_box

alt.HConcatChart(...)

In [186]:
charts_scatter = []

for column in reg_capped.columns:
    if column != 'grspaya':
        chart_s = alt.Chart(reg_capped).mark_point().encode(
            x=alt.X(column, title=column),
            y=alt.Y('grspaya', title='grspaya'),
            tooltip=[column, 'grspaya']
        ).properties(
            title=f'{column} vs grspaya'
        )
        charts_scatter.append(chart_s)

combined_scatter = alt.hconcat(*charts_scatter, spacing=20)

combined_scatter

alt.HConcatChart(...)

### Relationships

In [187]:
correlation_data = reg_corr.reset_index().melt('index')

base = alt.Chart(correlation_data).mark_rect().encode(
    x='index:N',
    y='variable:N',
    color='value:Q',
    tooltip=['index', 'variable', 'value']
)

highlight = base.mark_rect(stroke='red', strokeWidth=1).transform_filter(
    (alt.datum.index == 'grspaya') | (alt.datum.variable == 'grspaya')
)

heatmap = base + highlight

heatmap

alt.LayerChart(...)

The Variance Inflation Factor (VIF) analysis in a dataset can reveal significant insights into the extent of multicollinearity among independent variables in a linear regression model. This understanding is essential for ensuring the reliability and validity of the model’s outputs.

A VIF value close to 1 generally indicates minimal multicollinearity with other variables. Values between 1 and 5 are often seen as moderate and usually acceptable, though this can vary based on the specific analytical context. However, VIF values above 5 suggest the presence of potential multicollinearity issues, necessitating closer scrutiny. Particularly alarming are VIF values that exceed 10, as they strongly indicate significant multicollinearity. Such high levels of multicollinearity can inflate the variance and negatively impact the reliability of the regression coefficients.

In [188]:
vif_data = pd.DataFrame()
vif_data["feature"] = X_reg.columns
vif_data["VIF"] = [variance_inflation_factor(X_reg.values, i) for i in range(len(X_reg.columns))]
print(vif_data)

    feature        VIF
0     tvpol   5.342890
1   lrscale   7.884694
2    netuse   7.885694
3     tvtot   6.734677
4   nwsptot   2.689678
5   polintr  11.085126
6   mmbprty  47.346779
7    edude1  14.828831
8    edude3   3.973066
9   edufde1  10.043760
10  edufde3   5.455409
11  edumde1  10.763776
12  edumde3   2.427638


In certain datasets, variables such as tvpol, lrscale, netuse, tvtot, and edufde3 might show VIF values over 5, pointing towards possible multicollinearity concerns. Especially variables like lrscale and netuse approaching a threshold of 10 would be more concerning. Conversely, variables like polintr, mmbprty, edude1, edufde1, and edumde1 with VIF values significantly above 10 suggest a strong correlation with other variables in the model, potentially compromising the accuracy of the estimated coefficients.

Meanwhile, variables such as nwsptot, edude3, and edumde3, with relatively lower VIF values, indicate less concern regarding multicollinearity.

## Model

### Select model

### Training and validation

### Fit model

### Evaluation on test set

### Save model



Save your model in the folder `models/`. Use a meaningful name and a timestamp.

## Conclusions